src/coprocessor/dag/batch_executor/limit.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::interface::*;
+use crate::coprocessor::Result;
+use tipb::expression::FieldType;
+
+pub struct BatchLimitExecutor<Src: BatchExecutor> {
+    src: Src,
+    expect_rows: usize,
+}
+
+impl<Src: BatchExecutor> BatchLimitExecutor<Src> {
+    pub fn new(src: Src, limit: usize) -> Result<Self> {
+        if limit == 0 {
+            return Err(box_err!("limit should not be zero"));
+        }
+        Ok(Self {
+            src,
+            expect_rows: limit,
+        })
+    }
+}
+
+impl<Src: BatchExecutor> BatchExecutor for BatchLimitExecutor<Src> {
+    #[inline]
+    fn next_batch(&mut self, expect_rows: usize) -> BatchExecuteResult {
+        let mut result = self.src.next_batch(expect_rows);
+        if result.data.rows_len() <= self.expect_rows {
+            self.expect_rows -= result.data.rows_len();
+            return result;
+        }
+
+        result.data.truncate(self.expect_rows);
+        self.expect_rows = 0;
+        result.is_drained = Ok(true);
+        result
+    }
+
+    #[inline]
+    fn collect_statistics(&mut self, destination: &mut BatchExecuteStatistics) {
+        self.src.collect_statistics(destination);
+    }
+
+    #[inline]
+    fn schema(&self) -> &[FieldType] {
+        self.src.schema()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::coprocessor::codec::batch::{LazyBatchColumn, LazyBatchColumnVec};
+    use crate::coprocessor::codec::data_type::VectorValue;
+    use crate::coprocessor::dag::expr::EvalConfig;
+    use cop_datatype::{EvalType, FieldTypeAccessor, FieldTypeTp};
+    use tipb::expression::FieldType;
+
+    struct MockExecutor {
+        // ID(INT,PK), Foo(INT), Bar(Float,Default 4.5)
```

Comment:
Optional: maybe only 1 single column is enough for testing, because number of columns doesn't affect the result of limit. (Up to you to change)

Reply:
I prefer not to change since maybe we will use it in tests for other executors.

Reply:
I guess it may not be feasible enough for a fixed schema to be used in different executors. Finally it may be changed to something that needs more customization in different scenarios.